libexec/rtld-cheri-elf/riscv/rtld_machdep.h

```diff
@@ -28,8 +38,166 @@
  * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
+ *
+ * $FreeBSD$
  */
+
 #ifndef RTLD_MACHDEP_H
-#define RTLD_MACHDEP_H
-// TODO:
-#endif /* RTLD_MACHDEP_H */
+#define RTLD_MACHDEP_H	1
+
+__BEGIN_DECLS
+
+#include <sys/types.h>
+#include <cheri/cheri.h>
+#include <cheri/cheric.h>
+#include <machine/atomic.h>
+#include <stdlib.h>
+
+struct Struct_Obj_Entry;
+
+/* Return the address of the .dynamic section in the dynamic linker. */
+#define rtld_dynamic(obj) (&_DYNAMIC)
+
+Elf_Addr reloc_jmpslot(Elf_Addr *where, Elf_Addr target,
+    const struct Struct_Obj_Entry *defobj, const struct Struct_Obj_Entry *obj,
+    const Elf_Rel *rel);
+
+#define	FUNC_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	DATA_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_EXECUTE)
+
+/* TODO: we should have a separate member for .text/rodata */
+#define get_codesegment(obj) ((obj)->text_rodata_cap)
+
+/* TODO: ABIs with tight bounds */
+#define can_use_tight_pcc_bounds(obj) ((void)(obj), false)
+
+/*
+ * Create a pointer to a function.
+ * Important: this is not necessarily callable! For the PLT ABI we need a
+ * to load $cgp first -> use make_function_pointer() instead.
+ */
+static inline dlfunc_t
+make_code_pointer(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj,
+    bool tight_bounds, size_t addend)
+{
+	const void *ret = get_codesegment(defobj) + def->st_value;
+	/* Remove store and seal permissions */
+	ret = cheri_clearperm(ret, FUNC_PTR_REMOVE_PERMS);
+	if (tight_bounds) {
+		ret = cheri_setbounds(ret, def->st_size);
+	}
+	ret = cheri_incoffset(ret, addend); /* TODO: remove addend support */
```

Comment:
I can't remember why the addend was required, but I'm pretty sure I saw non-zero values...